### PR TITLE
select contiguous data splits for val and train

### PR DESCRIPTION
**Pull request by squassina**
Fixes:
- Data ablation study capability: allow skipping of sections
- select contiguous data splits for test and train

Includes changes for: 
- Update patch to patch_size in ablation.sh and README.md

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -142,11 +142,11 @@ To prepare the data for the experiments (e.g. split into train/val/test), please
 cd scripts
 
 # For section-based experiments
-python prepare_dutchf3.py split_train_val section --data-dir=${data_dir}/data
+python prepare_dutchf3.py split_train_val section --data_dir=${data_dir}/data
 
 
 # For patch-based experiments
-python prepare_dutchf3.py split_train_val patch --data-dir=${data_dir}/data --stride=50 --patch=100
+python prepare_dutchf3.py split_train_val patch --data_dir=${data_dir}/data --stride=50 --patch=100
```
Comment (squassina) [inline README.md]: Need to change patch=100 to patch_size=100

File: README.md [hunk 2]
```diff
@@ -296,8 +296,6 @@ This project has adopted the [Microsoft Open Source Code of Conduct](https://ope
 | **Core Tests** | master | [![Build Status](https://dev.azure.com/best-practices/deepseismic/_apis/build/status/microsoft.Tests%20(seismic-deeplearning)?branchName=master)](https://dev.azure.com/best-practices/deepseismic/_build/latest?definitionId=126&branchName=master) |
 | **Notebook Tests** | staging | [![Build Status](https://dev.azure.com/best-practices/deepseismic/_apis/build/status/microsoft.Notebooks%20(seismic-deeplearning)?branchName=staging)](https://dev.azure.com/best-practices/deepseismic/_build/latest?definitionId=125&branchName=staging) |
 | **Notebook Tests** | master | [![Build Status](https://dev.azure.com/best-practices/deepseismic/_apis/build/status/microsoft.Notebooks%20(seismic-deeplearning)?branchName=master)](https://dev.azure.com/best-practices/deepseismic/_build/latest?definitionId=125&branchName=master) |
-| **Azure ML Tests** | staging | TODO add badge link |
-| **Azure ML Tests** | master | TODO add badge link |
```
Comment (squassina) [inline README.md]: add the 2 lines back:
| **Azure ML Tests** | staging | TODO add badge link |	
| **Azure ML Tests** | master | TODO add badge link |

File: examples/interpretation/README.md [hunk 3]
```diff
@@ -1,8 +1,5 @@
...
-
+The repository contains the following notebook examples 
+* [Dutch F3 dataset](notebooks/F3_block_training_and_evaluation_local.ipynb): This notebook illustrates section and patch based segmentation approaches on the [Dutch F3](https://terranubis.com/datainfo/Netherlands-Offshore-F3-Block-Complete) open dataset.
 
 * [Penobscot dataset](notebooks/HRNet_Penobscot_demo_notebook.ipynb): 
-In this notebook, we demonstrate how to train an [HRNet](https://github.com/HRNet/HRNet-Semantic-Segmentation) model for facies prediction using [Penobscot](https://terranubis.com/datainfo/Penobscot) dataset. The Penobscot 3D seismic dataset was acquired in the Scotian shelf, offshore Nova Scotia, Canada. This notebook illustrates the use of HRNet based segmentation algorithm on the dataset. Details of HRNet based model can be found [here](https://arxiv.org/abs/1904.04514)
-
+This notebook illustrates the use of HRNet based segmentation algorithm on the [Penobscot](https://terranubis.com/datainfo/Penobscot) dataset
```
Comment (squassina) [inline examples/interpretation/README.md]: Review and keep contrib version

File: examples/interpretation/notebooks/F3_block_training_and_evaluation_local.ipynb [hunk 4]
```diff
@@ -262,8 +262,7 @@
     "# The number of epochs to run in training\n",
     "max_epochs = config.TRAIN.END_EPOCH  \n",
     "max_snapshots = config.TRAIN.SNAPSHOTS\n",
-    "dataset_root = config.DATASET.ROOT\n",
-    "model_pretrained = None"
+    "dataset_root = config.DATASET.ROOT"
```
Comment (squassina) [inline examples/interpretation/notebooks/F3_block_training_and_evaluation_local.ipynb]: Review and keep contrib version

File: examples/interpretation/notebooks/HRNet_Penobscot_demo_notebook.ipynb [hunk 5]
```diff
@@ -175,8 +175,7 @@
     "# The number of epochs to run in training\n",
     "max_epochs = config.TRAIN.END_EPOCH  \n",
     "max_snapshots = config.TRAIN.SNAPSHOTS\n",
-    "dataset_root = config.DATASET.ROOT\n",
-    "model_pretrained = config.MODEL.PRETRAINED"
+    "dataset_root = config.DATASET.ROOT"
```
Comment (squassina) [inline examples/interpretation/notebooks/HRNet_Penobscot_demo_notebook.ipynb]: Review and keep contrib version

File: tests/cicd/notebooks_build.yml [hunk 6]
```diff
@@ -5,16 +5,15 @@
 pr:
```
Comment (squassina) [inline tests/cicd/notebooks_build.yml]: Review and keep contrib version

File: tests/cicd/src/conftest.py [hunk 7]
```diff
@@ -7,7 +7,6 @@
 def pytest_addoption(parser):
     parser.addoption("--nbname", action="store", type=str, default=None)
     parser.addoption("--dataset_root", action="store", type=str, default=None)
-    parser.addoption("--model_pretrained", action="store", type=str, default=None)
```
Comment (squassina) [inline tests/cicd/src/conftest.py]: Review and keep contrib version

File: tests/cicd/src/notebook_integration_tests.py [hunk 8]
```diff
@@ -9,17 +9,11 @@
 # don't add any markup as this just runs any notebook which name is supplied
 # @pytest.mark.integration
 # @pytest.mark.notebooks
-def test_notebook_run(nbname, dataset_root, model_pretrained):
+def test_notebook_run(nbname, dataset_root):
```
Comment (squassina) [inline tests/cicd/src/notebook_integration_tests.py]: Review and keep contrib version

File: scripts/prepare_dutchf3.py [hunk 9]
```diff
@@ -105,66 +129,119 @@ def split_patch_train_val(data_dir, stride, patch_size, per_val=0.2, log_config=
...
+    horz_locations_test_begin = range(0, test_iline, max(1,stride))
+    test_i_list = list(_i_extract_patches(test_iline_range,
+                                           horz_locations_test_begin,
+                                           vert_locations))
+    # Process inlines - test - end
+    horz_locations_test_end = range(xline - stride + 1, xline, max(1,stride))
+    test_i_list += list(_i_extract_patches(test_iline_range,
+                                                horz_locations_test_end,
+                                                vert_locations))
+
```
[the remark this answers is not in the document]
Comment (squassina) [reply scripts/prepare_dutchf3.py]: All references to test is now val/validation

File: scripts/prepare_dutchf3.py [hunk 10]
```diff
@@ -9,49 +9,61 @@
...
-    file_object = open(path.join(splits_path, loader_type + "_train.txt"), "w")
+    file_object = open(path.join(splits_path,
+                       loader_type + "_train.txt"), "w")
     file_object.write("\n".join(train_list))
     file_object.close()
-    file_object = open(path.join(splits_path, loader_type + "_val.txt"), "w")
+    file_object = open(path.join(splits_path,
+                       loader_type + "_val.txt"), "w")
     file_object.write("\n".join(test_list))
     file_object.close()
```
[the remark this answers is not in the document]
Comment (squassina) [reply scripts/prepare_dutchf3.py]: All references to test is now val/validation

File: scripts/prepare_dutchf3.py [hunk 11]
```diff
@@ -105,66 +129,119 @@ def split_patch_train_val(data_dir, stride, patch_size, per_val=0.2, log_config=
...
-    logger.debug("Generating Inline patches")
-    logger.debug(horz_locations)
+    # Vertical locations is common to all patches processed
+    vert_locations = range(0, depth - patch_size, stride)
     logger.debug(vert_locations)
 
+    # Process inlines
+    # iline = xline x depth
+    test_iline = math.floor(xline * per_val / 2)
+    logger.debug(test_iline)
```
[the remark this answers is not in the document]
Comment (squassina) [reply scripts/prepare_dutchf3.py]: All references to test is now val/validation

File: tests/cicd/main_build.yml [hunk 12]
```diff
@@ -49,11 +49,16 @@ jobs:
...
-      echo "Starting unit tests"
+      echo "Starting cv_lib unit tests"
       source activate seismic-interpretation
       pytest --durations=0 cv_lib/tests/
-      echo "Unit test job passed"
-    
+      echo "cv_lib unit test job passed"
+      echo ""
+      echo "Starting scripts unit tests"
+      source activate seismic-interpretation
```
Comment (maxkazmsft) [inline tests/cicd/main_build.yml]: Why are we doing this twice?

File: scripts/prepare_dutchf3.py [hunk 13]
```diff
@@ -105,66 +128,118 @@ def split_patch_train_val(data_dir, stride, patch_size, per_val=0.2, log_config=
...
             for i, k in locations:
                 yield "x_" + str(i) + "_" + str(j) + "_" + str(k)
-    test_xline_range = list(test_xline_range)
-    test_x_list = list(_x_extract_patches(test_xline_range, horz_locations, vert_locations))
-    train_x_list = list(_x_extract_patches(train_xline_range, horz_locations, vert_locations))
+
+    logger.debug("Generating Crossline patches")
+    logger.debug("Generating Crossline patches - Train")
+    # xline = iline x depth
+    val_xline = math.floor(iline * per_val / 2)
```
Comment (yalaudah) [inline scripts/prepare_dutchf3.py]: why do we divide by 2 here? is it because the val sections are taken from both ends of the volume?

File: scripts/prepare_dutchf3.py [hunk 14]
```diff
@@ -105,66 +128,118 @@ def split_patch_train_val(data_dir, stride, patch_size, per_val=0.2, log_config=
...
-    vert_locations = range(0, depth - patch_size + 1, stride)
-    logger.debug("Generating Inline patches")
-    logger.debug(horz_locations)
+    # Vertical locations is common to all patches processed
+    vert_locations = range(0, depth - patch_size, stride)
     logger.debug(vert_locations)
 
+    # Process inlines
+    # iline = xline x depth
+    val_iline = math.floor(xline * per_val / 2)
```
Comment (yalaudah) [inline scripts/prepare_dutchf3.py]: why do we divide by 2 here? is it because the val sections are taken from both ends of the volume?

File: tests/cicd/main_build.yml [hunk 15]
```diff
@@ -49,11 +49,16 @@ jobs:
...
+      pytest --durations=0 tests/
```
Comment (maxkazmsft) [inline tests/cicd/main_build.yml]: Could you please place this in a separate build job? You can use cv_lib build job as an example, call this something like "job: common_tests" then include everything else: 
dependsOn: setup
timeoutInMinutes: 5
etc...

File: scripts/prepare_dutchf3.py [hunk 16]
```diff
@@ -105,66 +128,118 @@ def split_patch_train_val(data_dir, stride, patch_size, per_val=0.2, log_config=
...
-    train_x_list = list(_x_extract_patches(train_xline_range, horz_locations, vert_locations))
+
+    logger.debug("Generating Crossline patches")
+    logger.debug("Generating Crossline patches - Train")
+    # xline = iline x depth
+    val_xline = math.floor(iline * per_val / 2)
+    logger.debug(val_xline)
+
+    # Process xlines - train
+    horz_locations_train = range(val_xline, iline - patch_size, stride)
```
Comment (yalaudah) [inline scripts/prepare_dutchf3.py]: not sure i'm following here. Why is the range from val_xline to iline - patch_size? <-- it might be easier to jump on a quick call to discuss this.

File: scripts/prepare_dutchf3.py [hunk 17]
```diff
@@ -253,48 +328,89 @@ def split_alaudah_et_al_19(data_dir, stride, patch_size, fraction_validation=0.2
 
     list_train_val = i_list + x_list
 
-    # create train and test splits:
-    train_list, test_list = train_test_split(list_train_val, test_size=fraction_validation, shuffle=True)
+    # create train and validation splits:
+    train_list, val_list = train_val_split(list_train_val, val_size=fraction_validation, shuffle=True)
```
Comment (yalaudah) [inline scripts/prepare_dutchf3.py]: this is a function imported from scikit-learn. it should be `train_test_split` not `train_val_split`.. otherwise this would fail.

Comment (yalaudah) [reply scripts/prepare_dutchf3.py]: same goes for `val_size`.. it should be `test_size`

File: tests/test_prepare_dutchf3.py [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+        # reading the file and splitting the data
+        section_train = pd.read_csv(output + '/section_train.txt', header=None, names=['row'])
+        section_train = pd.DataFrame(section_train.row.str.split('_').tolist(), columns=['aline', 'section'])
+
+        section_val = pd.read_csv(output + '/section_val.txt', header=None, names=['row'])
+        section_val = pd.DataFrame(section_val.row.str.split('_').tolist(), columns=['aline', 'section'])
+
+        # test
+        assert (float(section_train.section[1]) - float(section_train.section[0])) % float(SLICE_STEPS) == 0.0
+        assert (float(section_val.section[1]) - float(section_val.section[0])) % float(SLICE_STEPS) > 0.0
```
Comment (yalaudah) [inline tests/test_prepare_dutchf3.py]: can you please explain why the the test requires the result to be 0 in the train case, and > 0 in the val case?

File: tests/cicd/main_build.yml (hunk 12 above)
Comment (squassina) [reply tests/cicd/main_build.yml]: it's 2 tests cv_lib and scripts. I added a new build job as per your other comment.

File: scripts/prepare_dutchf3.py (hunk 13 above)
Comment (squassina) [reply scripts/prepare_dutchf3.py]: yes, we split half to each edge of the volume.

File: scripts/prepare_dutchf3.py (hunk 14 above)
Comment (squassina) [reply scripts/prepare_dutchf3.py]: yes, we split half to each edge of the volume.

File: scripts/prepare_dutchf3.py (hunk 17 above)
Comment (squassina) [reply scripts/prepare_dutchf3.py]: I didn't catch this as an error because it is inside the split_alaudah_et_al_19 method. But I returned it to the original value.

File: tests/test_prepare_dutchf3.py (hunk 18 above)
Comment (squassina) [reply tests/test_prepare_dutchf3.py]: I only skip lines to the training set, and do nothing to validation set. That's why the test is like this.

This was as per the original issue opened [here](https://github.com/olgaliak/seismic-deeplearning/issues/9#issuecomment-582650229)

**Comment by yalaudah** [comment]
Issue #233 should also address this PR as well.
